sdks/python/apache_beam/io/tfrecordio_test.py

```diff
@@ -0,0 +1,365 @@
+
+  def setUp(self):
+    self.record = binascii.a2b_base64(FOO_RECORD_BASE64)
+
+  def _as_file_handle(self, contents):
+    result = cStringIO.StringIO()
+    result.write(contents)
+    result.reset()
+    return result
+
+  def _increment_value_at_index(self, value, index):
+    l = list(value)
+    l[index] = chr(ord(l[index]) + 1)
+    return ''.join(l)
+
+  def _test_error(self, record, error_text):
+    with self.assertRaises(ValueError) as context:
+      _TFRecordUtil.read_record(self._as_file_handle(record))
+    self.assertIn(error_text, context.exception.message)
+
+  def test_masked_crc32c(self):
+    self.assertEqual(0xfd7fffa, _TFRecordUtil._masked_crc32c('\x00' * 32))
+    self.assertEqual(0xf909b029, _TFRecordUtil._masked_crc32c('\xff' * 32))
+    self.assertEqual(0xfebe8a61, _TFRecordUtil._masked_crc32c('foo'))
+    self.assertEqual(
+        0xe4999b0,
+        _TFRecordUtil._masked_crc32c('\x03\x00\x00\x00\x00\x00\x00\x00'))
+
+  def test_masked_crc32c_crcmod(self):
+    crc32c_fn = crcmod.predefined.mkPredefinedCrcFun('crc-32c')
+    self.assertEqual(
+        0xfd7fffa,
+        _TFRecordUtil._masked_crc32c(
+            '\x00' * 32, crc32c_fn=crc32c_fn))
+    self.assertEqual(
+        0xf909b029,
+        _TFRecordUtil._masked_crc32c(
+            '\xff' * 32, crc32c_fn=crc32c_fn))
+    self.assertEqual(
+        0xfebe8a61, _TFRecordUtil._masked_crc32c(
+            'foo', crc32c_fn=crc32c_fn))
+    self.assertEqual(
+        0xe4999b0,
+        _TFRecordUtil._masked_crc32c(
+            '\x03\x00\x00\x00\x00\x00\x00\x00', crc32c_fn=crc32c_fn))
+
+  def test_write_record(self):
+    file_handle = cStringIO.StringIO()
+    _TFRecordUtil.write_record(file_handle, 'foo')
+    self.assertEqual(self.record, file_handle.getvalue())
+
+  def test_read_record(self):
+    actual = _TFRecordUtil.read_record(self._as_file_handle(self.record))
+    self.assertEqual('foo', actual)
+
+  def test_read_record_invalid_record(self):
+    self._test_error('bar', 'Not a valid TFRecord. Fewer than 12 bytes')
+
+  def test_read_record_invalid_length_mask(self):
+    record = self._increment_value_at_index(self.record, 9)
+    self._test_error(record, 'Mismatch of length mask')
+
+  def test_read_record_invalid_data_mask(self):
+    record = self._increment_value_at_index(self.record, 16)
+    self._test_error(record, 'Mismatch of data mask')
+
+  def test_compatibility_read_write(self):
+    for record in ['', 'blah', 'another blah']:
+      file_handle = cStringIO.StringIO()
+      _TFRecordUtil.write_record(file_handle, record)
+      file_handle.reset()
+      actual = _TFRecordUtil.read_record(file_handle)
+      self.assertEqual(record, actual)
+
+
+class TestTFRecordSink(unittest.TestCase):
+
+  def _write_lines(self, sink, path, lines):
+    f = sink.open(path)
+    for l in lines:
+      sink.write_record(f, l)
+    sink.close(f)
+
+  def test_write_record_single(self):
+    path = tempfile.NamedTemporaryFile().name
```

Comment:
What is happening here is that the file is created and then deleted on the first line. We are then using its name to re-create it later, and never deleting it. 

All of these tests should be changed to remove the temporary files they create. We could do this in tearDown, similar to how avroio_test.py works.